src/app/core/services/cart.service.ts

```diff
@@ -0,0 +1,89 @@
+import { Injectable } from '@angular/core';
+import { Observable, BehaviorSubject } from 'rxjs';
+
+import { DataService } from './data.service';
+import { IProduct } from 'src/app/shared/interfaces/product';
+
+import { environment } from '../../../environments/environment';
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class CartService {
+  private cart: IProduct[] = [];
+  private cartIds: String[] = [];
+  private products: IProduct[] = [];
+
+  private cartSource = new BehaviorSubject<IProduct[]>([]);
+
+  constructor(
+    private dataService: DataService,
+  ) {
+    const cartIds = JSON.parse(localStorage.getItem('cart'));
+    this.cartIds = cartIds ? cartIds : this.cartIds;
+  }
+
+  public addToCart(product: IProduct) {
+    product.addedToCart = true;
+    this.cart.push(product);
+    this.cartIds.push(product.id);
+    this.updateCart();
+  }
+
+  public removeFromCart(product: IProduct): void {
+    product.addedToCart = false;
+    const indexOfCurrId = this.cartIds.findIndex(el => el === product.id);
+    this.cartIds.splice(indexOfCurrId, 1);
+    const indexOfCurrProduct = this.cart.findIndex(el => el.id === product.id);
+    this.cart.splice(indexOfCurrProduct, 1);
+
+    this.updateCart();
+  }
+
+  public toggleCart(product: IProduct) {
+    if (!product.addedToCart) {
+      this.addToCart(product);
+      return;
+    }
+    this.removeFromCart(product);
+  }
+
+  private checkProductInCart(product: IProduct): IProduct {
```

Comment:
Functions with the name `check...` are usually expected to return a boolean (true/false). Or even better is to start the name with `is...` (e.g. you could name it `isAddedToCart `).
Also it's better when functions are pure (don't mutate their parameters)

Reply:
_"Also it's better when functions are pure (don't mutate their parameters)"_ 
- in our case mutation is conscious, as this function takes place as a product decorator